source/dol/bases/d_ac_py_key.cpp

```diff
@@ -0,0 +1,355 @@
+        mTriggeredButtons = mDownButtons & ~mPrevDownButtons;
+    } else if (mStatus & FLAG_FORCE_RIGHT) {
+        mDownButtons = BUTTON_RIGHT;
+        mTriggeredButtons = 0;
+    } else {
+        dGameKey_c *gameKey = dGameKey_c::m_instance;
+        mDownButtons = gameKey->keys[mRemoconID]->mDownButtons;
+        mTriggeredButtons = gameKey->keys[mRemoconID]->mTriggeredButtons;
+        if (mStatus & FLAG_DISABLE_LR) {
+            clearCrossKey();
+        }
+    }
+
+    if (mStatus & FLAG_SHAKE_COOLDOWN) {
+        offStatus(STATUS_SHAKE_COOLDOWN);
+        mShakeCooldown = 10;
+    }
+    if (mShakeCooldown != 0) {
+        mShakeCooldown--;
+    }
+
+    mActiveShakeJump = 0;
+    mTriggerShakeJump = 0;
+    if (!(mStatus & FLAG_NO_INPUT) && !(mStatus & FLAG_FORCE_RIGHT)) {
+        if (mStatus & FLAG_DEMO) {
+            mActiveShakeJump = mDemoDoShake;
+            mTriggerShakeJump = mDemoDoShake;
+        } else if (mShakeCooldown == 0 && dGameKey_c::m_instance->keys[mRemoconID]->mIsShaking) {
+            mActiveShakeJump = 1;
+            mTriggerShakeJump = 1;
+        }
+    }
+
+    if (!mHipAtkRead) {
+        mHipAtkHoldStart = false;
+        mHipAtkHoldCount = 0;
+    }
+    mHipAtkRead = false;
+
+    int *triggerHistory = mJumpTriggeredHistory[mRemoconID];
+    int *buttonHistory = mJumpButtonHistory[mRemoconID];
+    for (int i = ARRAY_SIZE(mJumpTriggeredHistory[0]) - 1; i >= 1; i--) {
+        triggerHistory[i] = triggerHistory[i - 1];
+        buttonHistory[i] = buttonHistory[i - 1];
+    }
+    triggerHistory[0] = triggerJumpBase();
+    buttonHistory[0] = buttonJump();
+    mTriggeredButtonsDemo = 0;
+    mDemoDoShake = 0;
+}
+
+void dAcPyKey_c::updateEnd() {
+    mActionTriggered = false;
+    if (mRemoconID == -1) {
+        return;
+    }
+    if (mStatus & FLAG_DEMO) {
+        return;
+    }
+    mDownButtons = dGameKey_c::m_instance->keys[mRemoconID]->mDownButtons;
+    mTriggeredButtons = dGameKey_c::m_instance->keys[mRemoconID]->mTriggeredButtons;
+}
+
+bool dAcPyKey_c::checkHipAttack() {
+    mHipAtkRead = true;
+    if (!mHipAtkHoldStart) {
+        if (triggerDown()) {
+            mHipAtkHoldStart = true;
+            mHipAtkHoldCount = 0;
+        }
+    } else if (buttonDown()) {
+        if (buttonHipAttack() && ++mHipAtkHoldCount >= 4) {
+            return true;
+        }
+    } else {
+        mHipAtkHoldStart = false;
+    }
+    return false;
+}
+
+void dAcPyKey_c::onStatus(u16 status) {
+    switch (status) {
+        case STATUS_FORCE_JUMP:
+            offStatus(STATUS_FORCE_NO_JUMP);
+            break;
+        case STATUS_FORCE_NO_JUMP:
+            offStatus(STATUS_FORCE_JUMP);
+            break;
+        case STATUS_NO_INPUT:
+        case STATUS_DEMO:
+        case STATUS_FORCE_RIGHT:
+            clearAllKey();
+            break;
+        case STATUS_DISABLE_LR:
+            clearCrossKey();
+            break;
+        default: break;
+    }
+    mStatus |= (1 << status);
+}
```

Comment:
I know this is pedantic but maybe some extra whitespace could make this more readable